src/helpers/nodeUtils.ts

```diff
@@ -223,57 +223,71 @@ export async function retrieveOrImportShare(params: {
     return Promise.reject(new Error(`invalid ${JSON.stringify(resultArr)}`));
   })
     .then((responses) => {
-      const promiseArrRequest: Promise<void | JRPCResponse<ShareRequestResult>>[] = [];
+      const promiseArrRequest: Promise<void | JRPCResponse<ShareRequestResult> | JRPCResponse<ShareRequestResult[]>>[] = [];
       const nodeSigs: CommitmentRequestResult[] = [];
       for (let i = 0; i < responses.length; i += 1) {
         const x = responses[i];
-        if (!x || typeof x !== "object") {
-          continue;
-        }
-        if (x.error) {
+        if (!x || typeof x !== "object" || x.error) {
           continue;
         }
         if (x) nodeSigs.push((x as JRPCResponse<CommitmentRequestResult>).result);
       }
-      for (let i = 0; i < endpoints.length; i += 1) {
-        const x = responses[i];
-        if (!x || typeof x !== "object") {
-          continue;
-        }
-        if (x.error) {
-          continue;
+
+      if (isImportShareReq) {
+        const importedPubKey = ecCurve.keyFromPublic({ x: importedShares[0].pub_key_x, y: importedShares[0].pub_key_x }).getPublic();
+        const importedPubKeyStr = importedPubKey.encode("hex", false).slice(2);
+        const hashedImportedPubKey = keccak256(Buffer.from(importedPubKeyStr, "utf8"));
+        const proxyEndpointNum = parseInt(hashedImportedPubKey, 16) % endpoints.length;
+        const sortedEndpoints = [endpoints[proxyEndpointNum], ...endpoints.slice(0, proxyEndpointNum), ...endpoints.slice(proxyEndpointNum + 1)];
```

Comment:
It is not secure to pass endpoints from frontend, malicious dapp can trick us to import it in a bad endpoint. Better is to get these endpoints on backend some how, may be from env variables or from node list.